### ISSUE TITLE
Feature Request: Replay attack prevention incremental counter

### ISSUE THREAD

**Issue by mrpops2ko** [OpenVPN/openvpn#353]
hi all, I was wondering if a modification could be on the table to better prevent unnecessary packet drops. 

I'll describe the issue first, which i'm sure probably a good 95% of openvpn users have experienced at one time or another. The dreaded replay attack prevention messages and subsequent dropped packets. 

The issue stems from a combination of packet queues in buffers filling up and the time based nature of openvpn's replay attack prevention. 

Lets say you are a residential user who has a modest connection of 200 mbps and the openvpn server is a fast pipe 1gb. You fire up that steam game download and o0nce you approach your max limits on speed, either as high upstream or downstream you'll find that gateway latencies go through the roof. This is a normal phenomena of nic buffer queues and we can mitigate that through the use of CAKE / FQ_CODEL and other solutions but those don't completely fix the issue, just help reduce the amount of packets that are erroneously dropped. 

Once that event occurs [high latency] we then start to see the openvpn gateways start to drop packets, thinking its a replay attack because the openvpn server hasn't received the packets in the time required (yes this is modifiable by some flags, but in practice i've found this to not resolve the issue for me and others).  The end result being that the packets then end up being erroneously dropped.

Another well known solution has the same replay attack prevention but it accomplishes it differently, through incremental counting instead of being time based. The end result is that when you are in the same exact scenario i've described - the packets just keep flowing even though the gateway latencies are high, rather than being dropped. 

Would this not just be a better scenario all round (incremental counting)? also please correct me if i have any misgivings on how the technologies work.

**Comment by schwabe** [comment]
I am not sure what you mean by "incremental counting". OpenVPN currently does not simply use only time based but also a window:

```
replay-window n [t]
Use a replay protection sliding-window of size n and a time window of t seconds.

By default n is 64 (the IPSec default) and t is 15 seconds.
```

15s is ages in Internet terms and only triggers if you receive a packet after one that has been sent more than 15s later than a packet with a higher id (sent later). And 64 packets is a quite substantial reordering. 

Also 15s at 200 MBit/s would be something like 1,6GB which is far far more than buffer space is available that could cause such a delay.

I think we need a better description of the issue you are running into to understand your problem and what a fix could be.

**Comment by mrpops2ko** [comment]
Ah I agree with most of your points in this.

I am making the point poorly, this is what I meant by incremental counting
```Nonces are never reused. A 64bit counter is used, and cannot be wound backward. UDP, however, sometimes delivers messages out of order. For that reason we use a sliding window, in which we keep track of the greatest counter received and a window of roughly 2000 prior values, checked after verifying the authentication tag. This avoids replay attacks while ensuring nonces are never reused and that UDP can maintain out-of-order delivery performance.```

The points you've made though suggest it isn't time related and I agree the values over those times would exceed the buffers. 

I'm not really sure how to describe the issue without just repeating all the observations I mentioned in the original post.

200mbps connection speed > large sequential download (steam game download for example) > monitor latency during this window (latency skyrockets) > openvpn logs start filling with replay attack messages > openvpn drops packets > loop of packet dropping and retransmission occurs, sometimes even bringing the gateway down from it. (not actually bringing the gateway down but causing a reconnection to the ovpn server)

Most people who i've talked to that do router based redirect gateway for all their traffic notice this kind of thing often. I've unfortunately migrated out of ovpn so i can't provide logs but its a very commonly googled topic. [here](https://forums.openvpn.net/viewtopic.php?t=26852#:~:text=11%3A53%20am-,When%20you%20get%20the%20replay%20error%20it%20usually%20means%20that,t%20suffer%20from%20extreme%20lag.&text=OpenVPN%20Inc.,-Posts%3A%201073%20Joined) and [here ](https://forums.openvpn.net/viewtopic.php?t=10905)and [here ](https://forum.netgate.com/topic/163238/mute-replay-warnings)and [here](https://forum.netgate.com/topic/87199/openvpn-keeps-restarting-authenticate-decrypt-packet-error)

Possibly I can't provide much more than 'look this happens and here is how to repeat it' I just kind of wondered how / why this happens and if it could be mitigated against. I tried increasing the window and playing about with the window time too and no matter what modifications I made it didn't seem to change anything. 

Some of the flags I was using were
```
client; persist-key; persist-tun; remote-cert-tls server; prng sha256 64; mlock; auth-nocache; replay-window 640 30
```

With the upcoming release of ovpn-DCO I was wondering if this issue will still be present and if there something im missing to mitigate it

**Comment by schwabe** [comment]
Yeah, it would be better to provide proper logs with information why OpenVPN drops/rejects these packets. With higher `--verb` OpenVPN will actually log the current stored window and so on. To get an idea what really happens with your connection.

What you describing as " incremental counting" is exactly what OpenVPN implements albeit with a default of 64 instead of 2000. If you set `replay-windows 2000` in your config, you get exactly what you are asking for.
